### PR TITLE
"Building multiple Progressive Web Apps on the same domain" article.

### LINKED ISSUE #5164: Building multiple Progressive Web Apps in the same domain [6/1/2021]
* [Draft](https://docs.google.com/document/d/1BMXcKV7UFxNkhSfSMY9Tulj21hu2pQ051Li0Yd4HVfA/edit)

### PR DESCRIPTION
**Pull request by demianrenzulli**
Fixes #5164

@jpmedley article is ready for review. We are still waiting for the assets of the other two authors before merging.

cc // @petele @kaycebasques

### CODE REVIEW

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:173]:
```suggestion
  for one app, it will wipe all the data from the origin; there's no way 
 to do this for a single app. Note that Chrome and some other
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:168]:
```suggestion
Here are some practical issues common to both same-origin approaches:
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:176]:
```suggestion
  well.
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:182]:
```suggestion
  feature.
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:190]:
```suggestion
you can't use a separate origin (e.g. a subdomain), as discussed in the [Using
separate origins](#using-separate-origins) section, from the two same-origin options we presented,
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:192]:
```suggestion
using non-overlapping paths is strongly recommended, over
overlapping/nested paths.
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:195]:
```suggestion
same-origin approaches. In the next section we'll go deeper into the details
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:196]:
```suggestion
of why using overlapping/nested paths is the least recommended strategy.
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:201]:
```suggestion
`https://example.com/` is the outer app and  `https://example.com/app/` is the
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:202]:
```suggestion
inner app), is that all URLs in the inner app will actually be considered
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:203]:
```suggestion
part of both the outer app and the inner app.
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:207]:
```suggestion
- **Installation Promotion:** If the user visits the inner app (for example,
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:208]:
```suggestion
  in a web browser), when the outer app is already installed in the user's
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:213]:
```suggestion
  that it is. The workaround for this is to install the inner app manually
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:214]:
```suggestion
  (via "Create Shortcut" browser menu option), or to install the inner app
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:215]:
```suggestion
  first, before the outer app.
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:217]:
```suggestion
  and the [Badging API](https://web.dev/badging-api/)**: If the outer app is
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:218]:
```suggestion
  installed but the inner app is not, notifications and badges coming from the
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:219]:
```suggestion
  inner app will be erroneously attributed to the outer app (which is the
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:224]:
```suggestion
  The outer app may capture URLs that belong to the inner app. This is
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:225]:
```suggestion
  especially likely if the outer app is installed but the inner app isn't.
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:226]:
```suggestion
  Similarly, links within the outer app that link to the inner app will not
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:227]:
```suggestion
  link capture into the inner app, since they are considered to be within the
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:228]:
```suggestion
  outer app's scope. Additionally, on Chrome OS and Android, if these apps are
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:231]:
```suggestion
  the outer app will capture all links. Even if the inner app is installed,
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:233]:
```suggestion
  the OS will still offer the user the choice of opening them in the outer
  app.
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:239]:
```suggestion
multiple Progressive Web Apps related to each other within the same **domain**.
```

Comment (jpmedley) [reply src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:239]: This is to be consistent with what you've done elsewhere.

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:243]:
```suggestion
many challenges, mainly because the browser won't fully consider
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:244]:
```suggestion
these to be distinct apps.
```

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (jpmedley) [inline src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:253]:
```suggestion
(for the outer app) and `https://example.com/app/` (for the inner app).
```

**Review by jpmedley — CHANGES_REQUESTED** [review]
Most of my requests are related to web.dev usage standards, though there are a few wording suggestions.

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (demianrenzulli) [reply src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:86]: Hi @jpmedley! That's what I wanted to say. There are site that have the home page at https://www.example.com (origin 1) and the checkout page at https://checkout.example.com (origin 2), and the login page at https://login.example.com (origin 3). This is actually a unique site, that, for some reason has been divided into multiple origins by using subdomains. If the developer wants to provide a single PWA experience across the site, that will have many challenges. That's why that alternative is not recommended.

Please, let me know if that's clear and if there's a better way of explaining it in the alternative image text.

File: src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md
Comment (demianrenzulli) [reply src/site/content/en/blog/building-multiple-pwas-on-the-same-domain/index.md:157]: I understand that in most cases we tend to talk about the "positive" case and avoid talking too much about the "negative" one and just say that it's not recommended. 
In this particular article, talking about the non-recommended use case is important because is a very common mistake developers do, and they don't really understand why it's a mistake.

Let me know if that makes sense.

cc // @mgiuca

**Review by jpmedley — APPROVED** [review]

**Comment by demianrenzulli** [comment]
Thanks for approving these changes, @jpmedley! We are still waiting for the profile details of one co-author, but if we don't receive them by today, I guess we can move forward and then add him at a later point.

I'll let you know when we are ready.

**Comment by demianrenzulli** [comment]
Hi @jpmedley! We just added the latest author, so we should be good to go now.

Two questions:

- I think we talked about this for other sites, but I'm not totally aware of the rules for Web.dev: should I merge directly?
- The article has an old date (date: 2021-05-18 attribute). I assume I should update this to the tentative date of publication? It seems to me that it would be good to get this published after the long weekend (Wednesday, June 2nd sounds like a good date).

Let me know what you think!

**Comment by jpmedley** [comment]
Do you have a squash and merge button that's not grayed out? If you do, you don't need my permission to use it.

To make it show up at a later date, put whatever date you want on it and add `scheduled: true` to the metadata. I think you'll need another approval from someone in DevRel, but I'm not sure about that. I've cc'd our contract tech writer since I'm about to be out of the office for several days.

cc: @rachelandrew

**Comment by demianrenzulli** [comment]
Thanks for your reply @jpmedley!

I can confirm I see the button in green, so I should be able to do the merge myself.
I just changed the publication date for next Wednesday, June 2nd and I added `scheduled: true` in the metadata.

@rachelandrew: please, let me know if you need to approve this as well, or if it's ok for me to proceed and merge.

Thanks!

**Comment by jpmedley** [comment]
I can confirm that she does not. We didn't mean to give you that impression. She's in a lot of discussions where her approval isn't required because as a contractor, her ability to self-select work is a little more limited than ours. We're just trying to keep her plate full.

**Comment by jpmedley** [comment]
We do have another problem though. "cmp" needs to be added to the authors file before this can merge. Do you need help with that?

**Comment by demianrenzulli** [comment]
Thanks @jpmedley!

Based on your response, I was about to merge this myself, but now I found that the linter is complaining that the recently created author "cmp" is missing.

@robdodson: my apologies if you are not the one to help on these cases. This article contains two new authors ("mgiuca" and "cmp"). As you can see both have been added to the `authorsData.json` file in this PR. 

I ran the linter before the latest commits I made and also checked out the PR locally once again and run the linter there and it works fine.

I'm wondering if the deployed version of the linter detects issues that the local version doesn't? 
Two ideas:

- The only special thing with the "cmp" author is that we haven't provided a profile image, since the author doesn't want to use one at this point. This field doesn't seem to be mandatory, but it might be causing some issues?
- The latest commit I did was changing the publication date for June 2nd and adding the `scheduled: true` attribute. Things seem to have failed after that.

We are very close to wrap this up, so thanks in advance for all the help.

**Comment by demianrenzulli** [comment]
@jpmedley @robdodson I can confirm that after running `npm run test` locally in the PR I was able to find the issue:

The tests complain about the missing author, but that works fine if the attribute is not present. It seems like the tests break when adding the `scheduled: true` attribute in the metadata (which was [my latest commit](https://github.com/GoogleChrome/web.dev/pull/5389/commits/b48244531973d29248f659d23af60abe4150db55)).

I'm wondering if this is a bug in web.dev? The use case seems to be adding a new author at the same time of creating an article with a late publication date and the `scheduled: true` attribute.

**Comment by robdodson** [comment]
@demianrenzulli I think I see what's going on. Assigning to @MichaelSolati to take a look. We should be able to get this fixed before the post's publish date.

**Comment by demianrenzulli** [comment]
@robdodson thanks for the prompt reply and let me know if you need me to test anything else on my side 👍

**Comment by MichaelSolati** [comment]
@demianrenzulli i think I fixed this just now, please try rebasing main on top of your branch and the tests should pass.

**Comment by robdodson** [comment]
@demianrenzulli I think you're correct, I wasn't able to get the tests to pass with that PR either. Apologies for the confusion! We can get it fixed up on Tuesday after folks are back from the holidays. I think that still leaves one more day before the due date.

**Comment by demianrenzulli** [comment]
Thanks for confirming @robdodson! By the way, Wednesday was a tentative (not a hard) deadline, so if it takes longer to get this fixed, we can always publish the following week.

**Comment by MichaelSolati** [comment]
Ok I think I misunderstood the issue and solved something completely irrelevant. I just opened #5484 which should address your issue, I believe it should... @robdodson will take a look at it on Tuesday I guess and then we can land it probably first thing on Tuesday as well.

**Comment by demianrenzulli** [comment]
Hi folks,

I assume that once the fix is done you would be able to merge this PR directly? The publishing date is supposed to be tomorrow, but just let me know if you need more time to fix it, so we can push it a few more days.

Cheers.

**Comment by MichaelSolati** [comment]
Not yet, just testing it, but should be ready soon we think.

**Comment by MichaelSolati** [comment]
@demianrenzulli ok, I believe that the `main` branch should now have the fix for your issue. I tested it locally. @robdodson I'm unsure if this will introduce any new issues, but I think we should be fine.

**Comment by demianrenzulli** [comment]
Thanks for the update @MichaelSolati!

This PR is blocked by the linter. If the changes are now in `main`, do you think that re-running the linter would help unblocking it? I'm not sure how often changes made in `main` end up hitting the servers where the linter runs.

**Comment by robdodson** [comment]
I think this branch would need to be rebased to fix the failing tests. @MichaelSolati are you able to do that?

**Comment by MichaelSolati** [comment]
Yes I can help, but can't at this second... I'm heading to the gym. @demianrenzulli do you think you'd be available in about 2 hours (4 PM PST)?

**Comment by demianrenzulli** [comment]
@MichaelSolati yes! just ping me later and I'll be able to test it and merge this PR.

**Comment by demianrenzulli** [comment]
Thanks for all the help @MichaelSolati! We tested the changes and they work well. Going to merge now. cc // @robdodson

**Pull request merged**
